src/main/java/io/smartcat/ranger/core/parser/DataGenerator.java

```diff
@@ -0,0 +1,128 @@
+        value.reset();
+        return value.get();
+    }
+
+    /**
+     * Builder for {@link DataGenerator}.
+     */
+    public static class Builder {
+
+        private final Map<String, Object> config;
+        private final String outputValueName;
+        private final Map<String, ValueProxy<?>> proxyValues;
+        private final ValueExpressionParser parser;
+        private final ReportingParseRunner<Value<?>> parseRunner;
+
+        /**
+         * Constructs Builder that will build {@link DataGenerator}.
+         *
+         * @param config Data generator configuration.
+         * @param outputValueName Name of output value.
+         */
+        public Builder(Map<String, Object> config, String outputValueName) {
+            this.config = config;
+            this.outputValueName = outputValueName;
+            this.proxyValues = new HashMap<>();
+            this.parser = Parboiled.createParser(ValueExpressionParser.class, proxyValues);
+            this.parseRunner = new ReportingParseRunner<>(parser.value());
+        }
+
+        /**
+         * Builds {@link DataGenerator} based on provided configuration.
+         *
+         * @return Instance of {@link DataGenerator}.
+         */
+        public DataGenerator build() {
+            createProxies();
+            parseCompositeValue("", config);
+            return new DataGenerator(proxyValues.get(outputValueName));
+        }
+
+        private void createProxies() {
+            for (Map.Entry<String, Object> entry : config.entrySet()) {
+                createProxy(entry.getKey(), entry.getValue());
+            }
+        }
+
+        @SuppressWarnings("unchecked")
+        private void createProxy(String key, Object value) {
+            proxyValues.put(key, new ValueProxy<>());
+            if (value instanceof Map) {
+                for (Map.Entry<String, Object> entry : ((Map<String, Object>) value).entrySet()) {
+                    createProxy(key + "." + entry.getKey(), entry.getValue());
+                }
+            }
+        }
+
+        @SuppressWarnings("unchecked")
+        private Value<?> parse(String parentName, Object def) {
+            if (def instanceof Map) {
+                return parseCompositeValue(parentName, (Map<String, Object>) def);
+            } else {
+                return parseSimpleValue(parentName, def);
+            }
+        }
+
+        @SuppressWarnings({ "rawtypes", "unchecked" })
+        private Value<?> parseCompositeValue(String parentName, Map<String, Object> def) {
+            Map<String, Value<?>> values = new HashMap<>();
+            for (String property : def.keySet()) {
+                String fullName = parentName.isEmpty() ? property : parentName + "." + property;
+                Value<?> val = parse(fullName, def.get(property));
+                ValueProxy proxy = proxyValues.get(fullName);
+                proxy.setDelegate(val);
+                values.put(property, proxy);
```

Comment:
```
ValueProxy proxy = new ValueProxy(val);
proxyValues.put(fullName, proxy);
values.put(property, proxy);
```

This will do same thing and we would avoid fetching value proxy and we can loose setter from ValueProxy. Am I missing something? Or you want proxyValues to hold ProxyValue without delegate and values to hold ProxyValue with delegate?